rust/kernel/print.rs

```diff
@@ -0,0 +1,775 @@
+/// to [`printk`]).
+///
+/// [`printk`]: ../../../../include/linux/printk.h
+#[doc(hidden)]
+pub unsafe fn call_printk(
+    format_string: &[u8; format_strings::LENGTH],
+    module_name: &[u8],
+    string: &[u8],
+) {
+    // `printk` does not seem to fail in any path.
+    bindings::printk(
+        format_string.as_ptr() as _,
+        module_name.len() as c_int,
+        module_name.as_ptr(),
+        string.len() as c_int,
+        string.as_ptr(),
+    );
+}
+
+/// Prints a message via the kernel's [`printk`] for the `CONT` "level".
+///
+/// Follows the interface and qualifiers of [`call_printk`] to make things
+/// easier in [`print_macro!`].
+///
+/// Hidden because it is called from a public macro -- it should never be
+/// called by users.
+///
+/// # Safety
+///
+/// See [`call_printk`]'s safety contract.
+///
+/// [`printk`]: ../../../../include/linux/printk.h
+#[doc(hidden)]
+pub unsafe fn call_printk_cont(
+    format_string: &[u8; format_strings::LENGTH_CONT],
+    _module_name: &[u8],
+    string: &[u8],
+) {
+    // `printk` does not seem to fail in any path.
+    bindings::printk(
+        format_string.as_ptr() as _,
+        string.len() as c_int,
+        string.as_ptr(),
+    );
+}
+
+// From `kernel/print/printk.c`.
+const LOG_LINE_MAX: usize = 1024 - 32;
+
+#[doc(hidden)]
+pub struct LogLineWriter {
+    data: [u8; LOG_LINE_MAX],
+    pos: usize,
+}
+
+impl LogLineWriter {
+    /// Creates a new [`LogLineWriter`].
+    pub fn new() -> LogLineWriter {
+        LogLineWriter {
+            data: [0u8; LOG_LINE_MAX],
+            pos: 0,
+        }
+    }
+
+    /// Returns the internal buffer as a byte slice.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.data[..self.pos]
+    }
+}
+
+impl Default for LogLineWriter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl fmt::Write for LogLineWriter {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        let copy_len = cmp::min(LOG_LINE_MAX - self.pos, s.as_bytes().len());
+        self.data[self.pos..self.pos + copy_len].copy_from_slice(&s.as_bytes()[..copy_len]);
+        self.pos += copy_len;
+        Ok(())
+    }
+}
+
+/// Performs formatting and forwards the string to [`call_printk`].
+///
+/// This is an *unsafe* macro.
+///
+/// Hidden because it is called from a public macro -- it should never be
+/// called by users.
+///
+/// # Safety
+///
+/// See [`call_printk`]'s and [`call_printk_cont`]'s safety contract.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! print_macro (
+    // Without extra arguments: no need to format anything, so simplify.
+    ($format_string:path, $call_fn:ident, $suffix:literal, $fmt:expr) => (
+        // SAFETY: This hidden macro should only be called by the documented
+        // printing macros which ensure the format string is a fixed one.
+        unsafe {
+            $crate::print::$call_fn(
+                &$format_string,
+                __MODULE_NAME,
```

Comment:
This won't work if someone splits their module into two files right?

Reply:
Ah, good point. What is the best option, `crate::`?

Reply:
I think so -- technically nothing requires that the `kernel_module!` macro be invoked in the root of a crate, but I think that'll work 99% of the time.

Reply:
Done in upcoming push.